### PR TITLE
UriUtils fixes

### PR DESCRIPTION
**Pull request by eduardb**
### Problem

Using `UriUtils#from(Uri)` for a second different `Uri`s will disrupt the first since the mapped IDs were held in a static field of the class. 

Also, numerical segments with value `0` weren't added to the mappings.

### Solution

`UriUtils#from(Uri)` shouldn't return an `UriUtils` instance in the first place, since it's a utility class with static methods.  So instead, we can have another static method that directly returns the mapping, that is using a local variable. In this way, several `Uri`s can be safely mapped at the same time 🎉 

### Tests

Updated existing tests for the mappings, and also include testing the mapping of an `Uri` that has a segment of value `0`.

### CODE REVIEW

File: core/src/main/java/novoda/lib/sqliteprovider/util/UriUtils.java [hunk 1]
```diff
@@ -6,29 +6,28 @@
...
+        Map<String, String> mappedIds = new HashMap<>();
 
         final List<String> segs = uri.getPathSegments();
         String parent = "";
 
         for (int i = 0; i < segs.size(); i++) {
             String currentSeg = segs.get(i);
             if (isNumeric(currentSeg)) {
-                final int children = Integer.parseInt(currentSeg);
-                for (int l = 0; l < children; l++) {
```
Comment (eduardb) [inline core/src/main/java/novoda/lib/sqliteprovider/util/UriUtils.java]: This `for` basically excluded `0` value children. But I don't see the logic behind that. If there is one, please someone point it out to me 😂

File: core/src/main/java/novoda/lib/sqliteprovider/util/UriUtils.java [hunk 2]
```diff
@@ -6,29 +6,28 @@
 
 public class UriUtils {
 
-    private static Map<String, String> mappedIds = new HashMap<String, String>();
+    private UriUtils() throws IllegalAccessException {
```
Comment (rock3r) [inline core/src/main/java/novoda/lib/sqliteprovider/util/UriUtils.java]: Make `UriUtils` `final`!

File: core/src/main/java/novoda/lib/sqliteprovider/util/UriUtils.java [hunk 3]
```diff
@@ -6,29 +6,28 @@
 
 public class UriUtils {
```
Comment (rock3r) [inline core/src/main/java/novoda/lib/sqliteprovider/util/UriUtils.java]: These changes require a major release ⚠️

File: core/src/main/java/novoda/lib/sqliteprovider/util/UriUtils.java [hunk 4]
```diff
@@ -6,29 +6,28 @@
...
 
-    private static Map<String, String> mappedIds = new HashMap<String, String>();
+    private UriUtils() throws IllegalAccessException {
+        throw new IllegalAccessException("Utils class shouldn't be instantiated");
```
Comment (rock3r) [inline core/src/main/java/novoda/lib/sqliteprovider/util/UriUtils.java]: Why throwing? It's also the wrong exception:

>An IllegalAccessException is thrown when an application tries to reflectively create an instance (other than an array), set or get a field, or invoke a method, but the currently executing method does not have access to the definition of the specified class, field, method or constructor.

This is not reflective. It should throw `InvalidOperationException` or something similar, if anything, but I'd simply do nothing in there.

**Review by rock3r — CHANGES_REQUESTED** [review]
Ok with the idea, but needs some changes

File: core/src/main/java/novoda/lib/sqliteprovider/util/UriUtils.java (hunk 3 above)
Comment (jackSzm) [inline core/src/main/java/novoda/lib/sqliteprovider/util/UriUtils.java]: Since we have a private c'tor , class can be final ?

File: core/src/main/java/novoda/lib/sqliteprovider/util/UriUtils.java [hunk 5]
```diff
@@ -6,29 +6,28 @@
...
 
-    public static UriUtils from(Uri uri) {
-        UriUtils utils = new UriUtils();
+    public static Map<String, String> mapIds(Uri uri) {
+        Map<String, String> mappedIds = new HashMap<>();
 
         final List<String> segs = uri.getPathSegments();
         String parent = "";
 
         for (int i = 0; i < segs.size(); i++) {
```
Comment (jackSzm) [inline core/src/main/java/novoda/lib/sqliteprovider/util/UriUtils.java]: Do we need a `fori` loop or just `for(String currentSeg : segs)` would be enough?

File: core/src/main/java/novoda/lib/sqliteprovider/util/UriUtils.java (hunk 1 above)
Comment (jackSzm) [inline core/src/main/java/novoda/lib/sqliteprovider/util/UriUtils.java]: Yeah this was excluding `0` and adding same value to the map `children` number of times ... 🤔

File: core/src/main/java/novoda/lib/sqliteprovider/util/UriUtils.java (hunk 4 above)
Comment (eduardb) [inline core/src/main/java/novoda/lib/sqliteprovider/util/UriUtils.java]: My train of thought was that someone might try to instantiate the class reflectively (don't ask me why). But I agree, simply not doing anything would be cleaner 😁

File: core/src/main/java/novoda/lib/sqliteprovider/util/UriUtils.java (hunk 5 above)
Comment (eduardb) [inline core/src/main/java/novoda/lib/sqliteprovider/util/UriUtils.java]: This is a small optimization to avoid creating an `Iterator` even when the list is empty

**Review by jszmltr — APPROVED** [review]

**Review by tasomaniac — APPROVED** [review]
Looks good to me. :+1:

File: core/src/main/java/novoda/lib/sqliteprovider/util/UriUtils.java (hunk 4 above)
Comment (eduardb) [inline core/src/main/java/novoda/lib/sqliteprovider/util/UriUtils.java]: Done in 84a86c2

File: core/src/main/java/novoda/lib/sqliteprovider/util/UriUtils.java (hunk 2 above)
Comment (eduardb) [inline core/src/main/java/novoda/lib/sqliteprovider/util/UriUtils.java]: Done in 84a86c2

File: core/src/main/java/novoda/lib/sqliteprovider/util/UriUtils.java (hunk 5 above)
Comment (florianmski) [inline core/src/main/java/novoda/lib/sqliteprovider/util/UriUtils.java]: Ugh, I'm all for readability over non critical optimizations, just my 2 cents

File: core/src/main/java/novoda/lib/sqliteprovider/util/UriUtils.java (hunk 1 above)
Comment (florianmski) [inline core/src/main/java/novoda/lib/sqliteprovider/util/UriUtils.java]: @devisnik might have the answer

Comment (eduardb) [inline core/src/main/java/novoda/lib/sqliteprovider/util/UriUtils.java]: @devisnik we summon you! 😄

File: core/src/main/java/novoda/lib/sqliteprovider/util/UriUtils.java (hunk 5 above)
Comment (eduardb) [inline core/src/main/java/novoda/lib/sqliteprovider/util/UriUtils.java]: I feel you @florianmski, but either way, it's out of the scope of this PR 😛

File: core/src/main/java/novoda/lib/sqliteprovider/util/UriUtils.java (hunk 1 above)
Comment (devisnik) [inline core/src/main/java/novoda/lib/sqliteprovider/util/UriUtils.java]: Sorry, guys, no idea. 

It's also excluding negative values, not just `0` though. So we might introduce some behavioural change here. Probably not an issue though. We might check on the usage side again though to learn what kind of Uris to expect here. Would be great to have that documented.

Disclaimer: I wasn't involved in any development of this library.

File: core/src/main/java/novoda/lib/sqliteprovider/util/UriUtils.java (hunk 5 above)
Comment (tasomaniac) [inline core/src/main/java/novoda/lib/sqliteprovider/util/UriUtils.java]: Since this will potentially run on very small List of Strings, it will not be a problem but actually this `fori` would potentially be slower because it does `segs.size()` on each loop. 
I agree that it is out of scope.

**Comment by friedger** [comment]
UriUtils is meant to help you with a uri, hence it makes sense to build UriUtils from a uri.

I would remove the static from mappedIds and I would add methods like getParentId(), hasParent(), isNumberedEntryWithinCollection(), etc.

Thereby it is possible to work on uris, like UriUtils.from("content://test.com/parent/1/child/6").getParent().getId() or so.

The class should be added to the extended demo. If there is no use case for UriUtils, then `from(..)` probably should be deprecated altogether.

**Review by rock3r — APPROVED** [review]

**Comment by eduardb** [comment]
Closing for the time being, as it still needs some small things to be addressed.
